packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,208 @@
+import {HTMLAttributes, MutableRefObject} from 'react';
+import {isElementVisible} from '@react-aria/focus/src/isElementVisible';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export interface AriaLandmarkProps  extends AriaLabelingProps {
+  role: 'main' | 'region' | 'search' | 'navigation' | 'form' | 'banner' | 'contentinfo' | 'complementary'
+}
+
+interface LandmarkAria {
+  landmarkProps: HTMLAttributes<HTMLElement>
+}
+
+let landmarkRoles = ['main', 'region', 'search', 'navigation', 'form', 'banner', 'contentinfo', 'complementary'];
+const LANDMARK_ELEMENT_SELECTOR = landmarkRoles.map(value => `[role=${value}]`).join(',');
+
+function getLandmarkTreeWalker(root: HTMLElement, opts?) {
+  let selector = LANDMARK_ELEMENT_SELECTOR;
+  let walker = document.createTreeWalker(
+    root,
+    NodeFilter.SHOW_ELEMENT,
+    {
+      acceptNode(node) {
+        // Not sure we need this, can landmarks be nested?
+        // Skip nodes inside the starting node.
+        // if (opts?.from?.contains(node)) {
+        //   return NodeFilter.FILTER_REJECT;
+        // }
+
+        if ((node as HTMLElement).matches(selector)
+          && isElementVisible(node as HTMLElement)
+          && opts?.registeredLandmarks.includes(node)) {
+          return NodeFilter.FILTER_ACCEPT;
+        }
+
+        return NodeFilter.FILTER_SKIP;
+      }
+    }
+  );
+
+  if (opts?.from) {
+    walker.currentNode = opts.from;
+  }
+
+  return walker;
+}
+
+function treeDistance(child: HTMLElement, target: HTMLElement): number {
+  let distance = 0;
+  let node = child;
+  while (target && node.parentElement && node !== target) {
+    node = node.parentElement;
+    distance++;
+  }
+  return distance;
+}
+
+class LandmarkManager {
+  private landmarks = new Map<MutableRefObject<HTMLElement>, HTMLElement>();
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  public addLandmark(ref: MutableRefObject<HTMLElement>) {
+    // could change to a map to track roles, then we could warn if someone provides two of the same role but no labels for them
+    if (!this.landmarks.has(ref)) {
+      this.landmarks.set(ref, null);
+    }
+  }
+
+  public removeLandmark(ref: MutableRefObject<HTMLElement>) {
+    if (this.landmarks.has(ref)) {
+      this.landmarks.delete(ref);
+    }
+  }
+
+  // Gets the landmark that is the closest parent in the DOM to the child
+  // Useful for nested Landmarks
+  private closestLandmark(child: HTMLElement) {
+    return Array.from(this.landmarks.keys())
+      .filter(landmark => (landmark?.current as HTMLElement).contains(child))
+      .sort((landmarkA, landmarkB) =>
+        treeDistance(child, landmarkA?.current) - treeDistance(child, landmarkB?.current)
+      )[0];
```

Comment:
Rather than sorting, couldn't you just walk up the tree from the until you find a node that is in the landmarks set? Should be linear time

Reply:
yep, makes perfect sense, over complicated that one

Reply:
Should be updated now